File: platform/frontend/lib/api/training.ts
/**
 * Training Job API
 */

import { apiClient } from "./client";

export interface TrainingJob {
  id: string;
  user_id: string;
  model_name: string;
  framework: string;
  dataset_s3_uri: string;
  config: Record<string, any> | null;
  status: "pending" | "running" | "completed" | "failed" | "cancelled";
  progress: number;
  message: string | null;
  checkpoint_s3_uri: string | null;
  error: string | null;
  created_at: string;
  updated_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface TrainingMetric {
  id: number;
  job_id: string;
  epoch: number;
  step: number | null;
  loss: number | null;
  accuracy: number | null;
  metrics: Record<string, any> | null;
  created_at: string;
}

export interface CreateTrainingJobRequest {
  user_id: string;
  model_name: string;
  framework: string;
  dataset_s3_uri: string;
  config?: Record<string, any>;
}

export const trainingApi = {
  async listJobs(params?: {
    user_id?: string;
    status_filter?: string;
    limit?: number;
    offset?: number;
  }): Promise<TrainingJob[]> {
    const response = await apiClient.get("/training/jobs", { params });
    return response.data;
  },

  async getJob(id: string): Promise<TrainingJob> {
    const response = await apiClient.get(`/training/jobs/${id}`);
    return response.data;
  },

  async createJob(data: CreateTrainingJobRequest): Promise<TrainingJob> {
    const response = await apiClient.post("/training/jobs", data);
    return response.data;
  },

  async cancelJob(id: string): Promise<void> {
    await apiClient.delete(`/training/jobs/${id}`);
  },

  async getMetrics(id: string, limit = 100): Promise<TrainingMetric[]> {
    const response = await apiClient.get(`/training/jobs/${id}/metrics`, {
      params: { limit },
    });
    return response.data;
  },
};

File: platform/backend/app/schemas/training.py
"""
Training API Schemas

Pydantic models for request/response validation.
"""

from datetime import datetime
from typing import Optional, Dict, Any
from uuid import UUID

from pydantic import BaseModel, Field

from app.db.models import JobStatus


# ==================== Request Schemas ====================


class TrainingJobCreate(BaseModel):
    """Request schema for creating a training job."""

    user_id: str = Field(..., min_length=1, max_length=255)
    model_name: str = Field(..., min_length=1, max_length=255, examples=["yolo11n"])
    framework: str = Field(..., min_length=1, max_length=100, examples=["ultralytics"])
    dataset_s3_uri: str = Field(
        ...,
        min_length=1,
        max_length=1024,
        examples=["s3://vision-platform/datasets/my-dataset/"],
    )
    config: Optional[Dict[str, Any]] = Field(
        default=None,
        examples=[{"epochs": 50, "batch_size": 16}],
    )


class TrainingUpdate(BaseModel):
    """Callback schema for training service to update job status."""

    job_id: str
    status: JobStatus
    progress: float = Field(ge=0.0, le=1.0, default=0.0)
    message: Optional[str] = None
    metrics: Optional[Dict[str, float]] = None  # e.g., {"epoch": 10, "loss": 0.5}
    checkpoint_s3_uri: Optional[str] = None
    error: Optional[str] = None


# ==================== Response Schemas ====================


class TrainingJobResponse(BaseModel):
    """Response schema for training job."""

    id: UUID
    user_id: str
    model_name: str
    framework: str
    dataset_s3_uri: str
    config: Optional[Dict[str, Any]]
    status: JobStatus
    progress: float
    message: Optional[str]
    checkpoint_s3_uri: Optional[str]
    error: Optional[str]
    created_at: datetime
    updated_at: datetime
    started_at: Optional[datetime]
    completed_at: Optional[datetime]

    model_config = {"from_attributes": True}


class TrainingMetricResponse(BaseModel):
    """Response schema for training metric."""

    id: int
    job_id: UUID
    epoch: int
    step: Optional[int]
    loss: Optional[float]
    accuracy: Optional[float]
    metrics: Optional[Dict[str, Any]]
    created_at: datetime

    model_config = {"from_attributes": True}
